### ISSUE TITLE
Upgrading dependent Golang runtime to 1.5

### ISSUE THREAD

**Issue by MNitta** [cloudfoundry/gorouter#113]
Upgrading dependent Golang runtime to 1.5
----------------------------------------

I was wondering if the CloudFoundry community planned to upgrade the gorouter dependent
Golang version to 1.5 or later to solve my problem. I am facing a strange problem that happens
sporadically.

That is, an application deployed on CloudFoundry returns 502 response, especially when many 
requests are sent simultaneously and the application is overloaded. When it happens, there are
error messages in the gorouter.log, something like

WARN 2016-01-07 16:41:44 :0: {"Error":"net/http: transport closed before response was received","Host":"xxxxxxx.xxxx.xxxx.net","Path":"/","RemoteAddr":"192.168.28.44:53822","RouteEndpoint":{"Addr":"192.168.20.78:62672","ApplicationId":"5a818a9d-175b-440c-9fdc-0ba01eb00a9f","Tags":{"component":"dea-4"}},"X-Forwarded-For":["111.111.111.11"],"X-Forwarded-Proto":["HTTPS"]} proxy.endpoint.failed WARN 2016-01-07 16:41:44 :0: {"Error":"net/http: transport closed before response was received","Host":"xxxxxxx.xxxx.xxxx.net","Path":"/","RemoteAddr":"192.168.28.44:53822","RouteEndpoint":{"Addr":"192.168.20.78:62672","ApplicationId":"5a818a9d-175b-440c-9fdc-0ba01eb00a9f","Tags":{"component":"dea-4"}},"X-Forwarded-For":["111.111.111.11"],"X-Forwarded-Proto":["HTTPS"]} 502 Bad Gateway: Registered endpoint failed to handle the request.

I guess these messages are from Golang runtime, specifically 
[https://github.com/golang/go/blob/release-branch.go1.4/src/net/http/transport.go#L1029][1]
and used in the roundtrip function. It seems to me that it waits for 100ms before closing the 
connection to avoid race condition.

This behavior has been improved by this commit:
[https://github.com/golang/go/commit/39377013cb9b49b39e55c83a5b221547f14d04dd#diff-6951e7593bfb1e773c9121df44df1c36][2]
and included in Golang version 1.5 (or maybe later).

I have concluded that my problem cannot be solved without upgrading golang version to 
1.5 or later. Even in the latest release of CloudFoundry, the golang version for gorouter is still 1.4.

  [1]: https://github.com/golang/go/blob/release-branch.go1.4/src/net/http/transport.go#L1029
  [2]: https://github.com/golang/go/commit/39377013cb9b49b39e55c83a5b221547f14d04dd#diff-6951e7593bfb1e773c9121df44df1c36

**Comment by ishustava** [comment]
@MNitta,

We are currently in the process of upgrading gorouter to golang 1.5.3. The upgrade is working its way through our pipeline.

Thanks for pointing out the issue.

**Comment by shalako** [comment]
Gorouter was upgraded to golang 1.5.3 in cf-release v230.

**Comment by shalako** [comment]
reopening for further investigation

**Comment by crhino** [comment]
Hi @MNitta, we were investigating whether the golang 1.5 upgrade in relation to your issue, testing out overloading an app of our own. When overloading our app with requests we saw this error occur even with 1.5.

[This comment](https://github.com/golang/go/blob/release-branch.go1.5/src/net/http/transport.go#L1193) in the `net/http` code highlights the two different cases where this error might occur. The race condition is one of those cases, but the other is the remote server hanging up prematurely, as it might when heavily overloaded.

We will close this issue, since we have updated to golang 1.5, but I would still expect to see these errors in the gorouter logs from time to time.
